workflows/cloudify_system_workflows/snapshot.py

```diff
@@ -0,0 +1,574 @@
+########
+# Copyright (c) 2015 GigaSpaces Technologies Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+
+
+import json
+import tempfile
+import shutil
+import zipfile
+import itertools
+import os
+import subprocess
+
+import elasticsearch
+import elasticsearch.helpers
+
+from datetime import datetime
+
+from cloudify import utils
+from cloudify.context import BootstrapContext
+from cloudify.decorators import workflow
+from cloudify.exceptions import NonRecoverableError
+from cloudify.manager import get_rest_client
+from cloudify_system_workflows.deployment_environment import \
+    generate_create_dep_tasks_graph
+
+
+_METADATA_FILE = 'metadata.json'
+
+# metadata fields
+_M_HAS_CLOUDIFY_EVENTS = 'has_cloudify_events'
+_M_VERSION = 'snapshot_version'
+
+_VERSION = '3.3'
+_AGENTS_FILE = 'agents.json'
+_ELASTICSEARCH = 'es_data'
+_CRED_DIR = 'credentials'
+_CRED_KEY_NAME = 'agent_key'
+_INFLUXDB = 'influxdb_data'
+_INFLUXDB_DUMP_CMD = ('curl -s -G "http://localhost:8086/db/cloudify/series'
+                      '?u=root&p=root&chunked=true" --data-urlencode'
+                      ' "q=select * from /.*/" > {0}')
+_INFLUXDB_RESTORE_CMD = ('cat {0} | while read -r line; do curl -X POST '
+                         '-d "[${{line}}]" "http://localhost:8086/db/cloudify/'
+                         'series?u=root&p=root" ;done')
+_STORAGE_INDEX_NAME = 'cloudify_storage'
+_EVENTS_INDEX_NAME = 'cloudify_events'
+
+
+class _DictToAttributes(object):
+    def __init__(self, dic):
+        self._dict = dic
+
+    def __getattr__(self, name):
+        return self._dict[name]
+
+
+def _get_json_objects(f):
+    def chunks(g):
+        ch = g.read(10000)
+        yield ch
+        while ch:
+            ch = g.read(10000)
+            yield ch
+
+    s = ''
+    n = 0
+    decoder = json.JSONDecoder()
+    for ch in chunks(f):
+        s += ch
+        try:
+            while s:
+                obj, idx = decoder.raw_decode(s)
+                n += 1
+                yield json.dumps(obj)
+                s = s[idx:]
+        except:
+            pass
+
+    assert not n or not s
```

Comment:
Perhaps raise an appropriate exception with a reasonable error message instead of this assertion?